qiita_db/artifact.py

```diff
@@ -0,0 +1,598 @@
+            generated. If not provided, `prep_template` should be provided.
+        processing_parameters : qiita_db.software.Parameter, optional
+            The processing parameters used to generate the new artifact
+            from `parents`. It is required if `parents` is provided. It should
+            not be provided if `prep_template` is provided.
+        can_be_submitted_to_ebi : bool, optional
+            Whether the new artifact can be submitted to EBI or not. Default:
+            `False`.
+        can_be_submitted_to_vamps : bool, optional
+            Whether the new artifact can be submitted to VAMPS or not. Default:
+            `False`.
+
+        Returns
+        -------
+        qiita_db.artifact.Artifact
+            A new instance of Artifact
+
+        Raises
+        ------
+        QiitaDBArtifactCreationError
+            If `filepaths` is not provided
+            If both `parents` and `prep_template` are provided
+            If none of `parents` and `prep_template` are provided
+            If `parents` is provided but `processing_parameters` is not
+            If both `prep_template` and `processing_parameters` is provided
+            If not all the artifacts in `parents` belong to the same study
+
+        Notes
+        -----
+        The visibility of the artifact is set by default to `sandbox`
+        """
+        # We need at least one file
+        if not filepaths:
+            raise QiitaDBArtifactCreationError(
+                "at least one filepath is required.")
+
+        # Parents or prep template must be provided, but not both
+        if parents and prep_template:
+            raise QiitaDBArtifactCreationError(
+                "parents or prep_template should be provided but not both")
+        elif not (parents or prep_template):
+            raise QiitaDBArtifactCreationError(
+                "at least parents or prep_template must be provided")
+
+        # If parents is provided, processing parameters should also be provided
+        if parents and not processing_parameters:
+            raise QiitaDBArtifactCreationError(
+                "if parents is provided, processing_parameters should also be"
+                "provided.")
+
+        # If prep_template is provided, processing_parameters should not be
+        # provided
+        if prep_template and processing_parameters:
+            raise QiitaDBArtifactCreationError(
+                "if prep_template is provided, processing_parameters should "
+                "not be provided.")
+
+        if timestamp is None:
+            timestamp = datetime.now()
+
+        with TRN:
+            visibility_id = convert_to_id("sandbox", "visibility")
+            artifact_type_id = convert_to_id(artifact_type, "artifact_type")
+
+            if parents:
+                # Check that all parents belong to the same study
+                studies = {p.study.id for p in parents}
```

Comment:
Currently all accesses to the DB are in a TRN (you can search for my previous PR a few weeks ago about transactionizing everything). That's why through the code there are `with TRN:` even if it looks like is not doing any SQL call.

I can put all validations in the same block of if/elifs but then all the checks will be done inside the TRN, in order to make sure that all the DB accesses performed in the `create` method are done in a Transaction. I think as it is right now is better, but if others have strong feelings I can transactionize everything.